packages/frontend/src/lib/pnl.ts

```diff
@@ -29,62 +34,141 @@ type ShortGainParams = {
+export function calcShortGain({ shortUnrealizedPNL, usdAmount, wethAmount, uniswapEthPrice }: ShortGainParams) {
+  if (
+    wethAmount.isEqualTo(0) ||
+    shortUnrealizedPNL.isEqualTo(0) ||
+    uniswapEthPrice.isEqualTo(0) ||
+    usdAmount.isEqualTo(0)
+  ) {
     return new BigNumber(0)
   }
-  return shortUnrealizedPNL.div(usdAmount.plus(wethAmount.times(ethPrice).absoluteValue())).times(100)
-}
-export function calcLongUnrealizedPnl({
-  sellQuote,
-  wethAmount,
-  ethPrice,
-}: {
-  sellQuote: BigNumber
-  wethAmount: BigNumber
-  ethPrice: BigNumber
-}) {
-  return {
-    usdValue: sellQuote.minus(wethAmount.abs()).multipliedBy(ethPrice),
-    ethValue: sellQuote.minus(wethAmount.abs()),
-  }
+  return shortUnrealizedPNL.div(usdAmount.plus(wethAmount.times(uniswapEthPrice).absoluteValue())).times(100)
 }
 
-export function calcETHCollateralPnl(
+export async function calcETHCollateralPnl(
   data: VaultHistory_vaultHistories[] | undefined,
-  ethPriceMap: { [key: string]: number },
-  ethPrice: BigNumber,
+  uniswapEthPrice: BigNumber,
+  currentVaultEthBalance: BigNumber,
 ) {
-  const deposits = data?.length
-    ? data?.reduce((acc: BigNumber, curr: VaultHistory_vaultHistories) => {
-        const time = new Date(Number(curr.timestamp) * 1000).setUTCHours(0, 0, 0) / 1000
-        if (curr.action === Action.DEPOSIT_COLLAT) {
-          acc = acc.plus(
-            new BigNumber(toTokenAmount(curr.ethCollateralAmount, 18)).times(
-              new BigNumber(ethPrice).minus(ethPriceMap[time]),
-            ),
-          )
-        }
-        return acc
-      }, new BigNumber(0))
-    : new BigNumber(0)
-
-  const withdrawals = data?.length
-    ? data?.reduce((acc: BigNumber, curr: VaultHistory_vaultHistories) => {
-        const time = new Date(Number(curr.timestamp) * 1000).setUTCHours(0, 0, 0) / 1000
-        if (curr.action === Action.WITHDRAW_COLLAT) {
-          acc = acc.plus(
-            new BigNumber(toTokenAmount(curr.ethCollateralAmount, 18)).times(
-              new BigNumber(ethPrice).minus(ethPriceMap[time]),
-            ),
+  const { deposits, withdrawals } = data?.length
+    ? await data?.reduce(async (prevPromise, vaultHistory: VaultHistory_vaultHistories) => {
+        const acc = await prevPromise
+
+        const ethPriceAtTimeOfAction = await getEthPriceAtTransactionTime(vaultHistory.timestamp)
+
+        if (vaultHistory.action === Action.DEPOSIT_COLLAT) {
+          acc.deposits = ethPriceAtTimeOfAction
+            ? acc.deposits.plus(
+                new BigNumber(toTokenAmount(vaultHistory.ethCollateralAmount, 18)).times(ethPriceAtTimeOfAction),
+              )
+            : BIG_ZERO
+        } else if (vaultHistory.action === Action.WITHDRAW_COLLAT) {
+          acc.withdrawals = acc.withdrawals.plus(
+            new BigNumber(toTokenAmount(vaultHistory.ethCollateralAmount, 18)).times(ethPriceAtTimeOfAction),
```

Comment:
Thinking that we need a ethPriceAtTimeOfAction condition check for withdrawals as well